File: python/cowgirl.py
# https://oj.vnoi.info/problem/cowgirl

mask = 3


def check_fn(a, b, n):
    for k in range(n - 1):
        if ((a >> (k)) & mask) | (((b >> (k)) & mask) << 2) in [0, 15]:
            return False
    # for i in range(n-1):
    #     if (a&(1<<i))!=0 and (a&(1<<(i+1)))!=0 and (b&(1<<i))!=0 and (b&(1<<(i+1)))!=0:return False
    #     if (a&(1<<i))==0 and (a&(1<<(i+1)))==0 and (b&(1<<i))==0 and (b&(1<<(i+1)))==0:return False

    return True


if __name__ == "__main__":
    tc = int(input())

    for t in range(tc):
        [m, n] = map(int, input().split(" "))
        if m < n:
            m, n = n, m
        check = [[0 for _ in range(1 << n)] for _ in range(1 << n)]
        f = [[0 for _ in range(1 << n)] for _ in range(m)]

        for i in range(1 << n):
            for j in range(1 << n):
                check[i][j] = check_fn(i, j, n)

        s = 0
        for i in range(m):
            for j in range(1 << n):
                if i == 0:
                    f[i][j] = 1
                else:
                    f[i][j] = sum([f[i - 1][k] for k in range(1 << n) if check[j][k]])
                if i == m - 1:
                    s += f[i][j]

        # print(sum(f[m-1][k] for k in range(1<<n)))
        print(s)

File: src/bin/cowgirl.rs
// const V: [usize; 2] = [0, 15];

fn check_fn(a: usize, b: usize, n: usize) -> bool {
    // for i in 0..(n - 1) {
    //     if V.contains(&(((a >> i) & 3) | (((b >> i) & 3) << 2))) {
    //         return false;
    //     }
    // }
    //
    // return true;

    let x = a & b;
    let y = (!a & !b) & ((1 << n) - 1);

    (x & x >> 1) == 0 && (y & y >> 1) == 0
}

fn main() {
    // read input from stdin
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).unwrap();

    // parse input to int
    let tc = input.trim().parse::<usize>().unwrap();
    for _ in 0..tc {
        // read input from stdin
        let mut input = String::new();
        std::io::stdin().read_line(&mut input).unwrap();
        // split input by space and parse to int, the input is a string in form of "a b"
        let mut iter = input.trim().split_whitespace();
        let mut m = iter.next().unwrap().parse::<usize>().unwrap();
        let mut n = iter.next().unwrap().parse::<usize>().unwrap();
        // swap m and n if m > n
        if m < n {
            (m, n) = (n, m);
        }

        let mut f = vec![vec![0; 1 << n as usize]; m as usize];
        let mut check = vec![vec![true; 1 << n as usize]; 1 << n as usize];
        for i in 0..(1 << n) {
            for j in (i)..(1 << n) {
                check[i][j] = check_fn(i, j, n);
            }
        }

        let mut sum = 0;

        for i in 0..m {
            for j in 0..(1 << n) {
                if i == 0 {
                    f[i][j] = 1;
                } else {
                    // compute f[i][j] = sum from 0 to (1<<n) of f[i-1][k] if check[j][k] is true
                    for k in 0..(1 << n) {
                        if check[j][k] && check[k][j] {
                            f[i][j] += f[i - 1][k];
                        }
                    }
                }
                if i == (m - 1) {
                    sum += f[i][j];
                }
            }
        }

        println!("{}", sum);
    }
}
